core/store/src/trie/mem/updating.rs

```diff
@@ -33,15 +71,100 @@ pub enum UpdatedMemTrieNode {
     },
     Extension {
         extension: Box<[u8]>,
-        child: OldOrUpdatedNodeId,
+        child: GenericNodeOrIndex<GenericTrieNodePtr>,
     },
     /// Corresponds to either a Branch or BranchWithValue node.
     Branch {
-        children: Box<[Option<OldOrUpdatedNodeId>; 16]>,
+        children: Box<[Option<GenericNodeOrIndex<GenericTrieNodePtr>>; 16]>,
         value: Option<FlatStateValue>,
     },
 }
 
+pub type UpdatedMemTrieNode = GenericUpdatedTrieNode<MemTrieNodeId, FlatStateValue>;
+
+/// An updated node with its memory usage.
+/// Needed to recompute subtree function (memory usage) on the fly.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct GenericUpdatedTrieNodeWithSize<GenericTrieNodePtr, FlatStateValue> {
+    pub node: GenericUpdatedTrieNode<GenericTrieNodePtr, FlatStateValue>,
+    pub memory_usage: u64,
+}
+
+impl<N, V> GenericUpdatedTrieNodeWithSize<N, V> {
+    pub fn empty() -> Self {
+        Self { node: GenericUpdatedTrieNode::Empty, memory_usage: 0 }
+    }
+}
+
+impl<GenericTrieNodePtr, FlatStateValue: HasLength>
+    GenericUpdatedTrieNode<GenericTrieNodePtr, FlatStateValue>
+{
+    fn memory_usage_value(value_length: u64) -> u64 {
+        value_length * TRIE_COSTS.byte_of_value + TRIE_COSTS.node_cost
+    }
+
+    pub fn memory_usage_direct(&self) -> u64 {
```

Comment:
Is this method meant to replace both `TrieNode::memory_usage_direct_internal` and `MemTrieNodeView::memory_usage`?